feat: Bulk Salary Structure Assignment

krantheman opened the pull request:
## Bulk Salary Structure Assignment

This PR introduces the **Bulk Salary Structure Assignment _DocType_**, a replacement and enhancement to the **Bulk Salary Structure Assignment _Dialog_**.

### The Problem - Bulk Salary Structure Assignment Dialog

![image](https://github.com/frappe/hrms/assets/61287991/238ab751-fbf7-4160-b6ac-929aa53d3b2d)

Issues with the BSSA Dialog include, but are not limited to:
- Inability to select specific Employees.
- Inability to set varying Base & Variable amounts for different Employees.
- Limited filtering options.
- Prerequisite navigation to a Salary Structure document.
- Potential loss of progress due to clicking outside.

### The Solution - Bulk Salary Structure Assignment DocType

https://github.com/frappe/hrms/assets/61287991/595a9688-2b41-4bd1-b60b-8d4efb405343

Tackling the aforementioned issues, features of the BSSA DocType include:
- Ability to select specific Employees.
- Ability to set varying Base & Variable amounts. This can also be done in bulk via row selection.
- Addition of the Filter Group component to set complex Employee filters.
- Ease of navigation and selection of any Salary Structure from within the page.
- Automatic fetching of Default Base Pay of Employees based on their Grade.
- Identification of Employees without preexisting Assignments for a specific date.

### Miscellaneous

In this PR, I have also:
- Edited the Create -> Bulk Assignments button in Salary Structure to redirect to the BSSA DocType.
- Added Bulk Salary Structure Assignment to the Salary Payout workspace and the Salary Structure List View.
- _Commonified_ the `notify_status` function by moving it from `leave_control_panely.py` to `utils.py`.
- _Commonified_ the `create_employee_grade` function by moving it from `test_salary_structure.py` to `test_utils.py`. 

### Documentation

https://frappehr.com/docs/v14/en/salary-structure-assignment-tool

krantheman committed:
feat: create Bulk Salary Structure Assignment doctype

krantheman committed:
feat: add Quick and Advanced Filters

krantheman committed:
refactor: move Base & Variable to datatable

krantheman committed:
refactor: redirect Bulk Assignments in Salary Structure to Bulk Salary Structure Assignment

krantheman committed:
feat: add Bulk Assign server side functionality

krantheman committed:
feat: status notification

krantheman committed:
feat: add number validation for base & variable

krantheman committed:
feat: add Bulk Salary Structure Assignment to Salary Payout workspace

krantheman committed:
fix: user perms

krantheman committed:
feat: add update Base/Variable for multiple rows

krantheman committed:
feat: load default base pay based on employee grade

krantheman committed:
fix: use updated values while assigning structure

krantheman committed:
feat: add Employee Grade column

krantheman committed:
feat: validate selection of employees while setting Base/Variable

krantheman committed:
refactor: render update button only on row selection

krantheman committed:
chore: improve readability

krantheman committed:
feat: prevent loading of employees without From Date

krantheman committed:
test: add test_get_employees

krantheman committed:
chore: move create_employee_grade to test_utils

krantheman committed:
test: add test_bulk_assign_structure

krantheman committed:
chore: fix formatting

ruchamahabal commented on hrms/hr/utils.py:
```diff
@@ -773,3 +774,42 @@ def get_ec_matching_query(bank_account, company, exact_match, from_date=None, to
 	"""
+
+
+def notify_status(doctype: str, failure: list, success: list) -> None:
```
should we call it `notify_bulk_action_status` now that it's generic?

ruchamahabal commented on hrms/payroll/doctype/salary_structure/salary_structure.js:
```diff
@@ -104,7 +104,10 @@ frappe.ui.form.on("Salary Structure", {
 
 			frm.add_custom_button(__("Bulk Assignments"), () => {
-				frm.trigger("assign_to_employees")
+				const doc = frappe.model.get_new_doc("Bulk Salary Structure Assignment");
```
Let's add this button to the Salary Structure list view too

ruchamahabal commented on hrms/payroll/doctype/salary_structure/salary_structure.py:
```diff
@@ -174,128 +174,56 @@ def get_employees(self, **kwargs):
 		return employees
 
-	@frappe.whitelist()
-	def assign_salary_structure(
```
Can't really break a public function in a minor release 🙈. Might break people's implementations relying on this API. Let's keep this and call our function from here, if possible?

ruchamahabal commented on hrms/payroll/doctype/bulk_salary_structure_assignment/bulk_salary_structure_assignment.py:
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from pypika import functions as fn
```
```suggestion
from pypika.functions import Coalesce
```
You can directly import Coalesce

ruchamahabal commented on hrms/payroll/doctype/bulk_salary_structure_assignment/bulk_salary_structure_assignment.py:
```diff
@@ -0,0 +1,106 @@
+			.on(Employee.grade == Grade.name)
+			.select(
+				fn.Coalesce(Grade.default_base_pay, 0).as_("base"),
+				fn.Coalesce(0).as_("variable"),
```
You can use `ConstantColumn` instead
https://frappeframework.com/docs/user/en/api/query-builder#constant-column

ruchamahabal commented on hrms/payroll/doctype/bulk_salary_structure_assignment/bulk_salary_structure_assignment.py:
```diff
@@ -0,0 +1,106 @@
+		if len(employees) <= 20:
+			return _bulk_assign_structure()
+
+		return frappe.enqueue(_bulk_assign_structure, timeout=600)
```
```suggestion
		return frappe.enqueue(_bulk_assign_structure, timeout=3000)
```
600 would be too low given that it's a bulk assignment tool. Maybe 3000 like payroll entry? Long queue has a timeout of 1500 so anything roughly between 1500 - 3000

ruchamahabal commented on hrms/payroll/doctype/bulk_salary_structure_assignment/bulk_salary_structure_assignment.js:
```diff
@@ -0,0 +1,325 @@
+
+	set_primary_action(frm) {
+		frm.disable_save();
+		frm.page.set_primary_action(__("Assign Structure"), () => {
```
To make the UX even better, we can also warn about validations beforehand. Eg: 
- Base is required almost every time. Just warn the user that you have not set the base for these rows. Want to proceed?
- Also `validate_dates` function in salary structure assignment checks if we are trying to assign a structure before DOJ or after relieving. Maybe flag such rows - color-code/gray out with a hover message on why structure cannot be assigned to them. Or other option is don't fetch such employees at all

ruchamahabal commented:
Very clean implementation overall 👏

krantheman committed:
chore: rename notify_status function

krantheman committed:
feat: add Bulk Salary Structure Assignment button to Salary Structure List View

krantheman committed:
chore: use ConstantColumn instead of Coalesce

krantheman committed:
fix: queue timeout

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

krantheman committed:
feat: exclude unjoined and relieved employees while fetching employees

krantheman committed:
chore: better error msg

krantheman committed:
feat: validate employees with base 0

krantheman committed:
revert: bring back public assign_salary_structure function

ruchamahabal committed:
Merge branch 'develop' into bulk-salary-structure-assignment

ruchamahabal committed:
chore: fix message

ruchamahabal committed:
refactor: fix transactional assignment creation

- move savepoint inside the for loop

- let calling function handle errors and logging

- fix logic for publishing progress - gets stuck if a failure occurs

ruchamahabal commented:
enqueue function should be accessible outside the class (public function). Also a message to inform the user that its enqueued

![image](https://github.com/frappe/hrms/assets/24353136/d2a41f8b-3426-43f8-bc7b-d3880dad5022)

Can you make this change? Just test it once by enqueuing irrespective of the number. Also lets enqueue if its > 30

krantheman committed:
fix: enqueue function call

krantheman committed:
chore: use optional chaining while checking for success response

krantheman committed:
fix: publish_progress in case of queue

krantheman committed:
fix: notify status in case of enqueued job

krantheman committed:
chore: improve readability

krantheman committed:
feat: add confirmation dialog while assigning structure

ruchamahabal committed:
Merge branch 'develop' into bulk-salary-structure-assignment

ruchamahabal commented:
ERPNext install is failing on develop. Merging this as its unrelated and tests are passing locally
